src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs

```diff
@@ -684,7 +703,7 @@ private void ProduceStart(bool appCompleted)
 
         protected Task TryProduceInvalidRequestResponse()
         {
-            if (_requestRejected)
+            if (_requestRejected || _applicationException != null)
```

Comment:
Indeed. My very first change (that I didn't even commit) was calling `Abort()` on a content length mismatch. This would prevent the call to `ProduceEnd()` in `FrameOfT`. But since I moved away from that decision, `ProduceEnd()` handles it as expected.